### ISSUE TITLE
get_SDA_property() results all NA with some horizon properties

### ISSUE THREAD

**Issue by dylanbeaudette** [ncss-tech/soilDB#261]
I'm not sure when this happened, but seeing a lot of `NA` returned by `get_SDA_property()` for map unit keys / components / horizons that should return valid results. Only happens when including some horizon properties such as `wthirdbar_r`.

Example:
```r
library(soilDB)

v <- c("sandtotal_r", "silttotal_r", "claytotal_r", "om_r", "cec7_r", "ph1to1h2o_r", "wthirdbar_r", "wtenthbar_r", "wfifteenbar_r")

# source data
get_SDA_property(
  property = v,
  method = "none", 
  mukeys = 341081,
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)

# all NA
get_SDA_property(
  property = v,
  method = "Weighted Average", 
  mukeys = 341081,
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)

# works
get_SDA_property(
  property = v[1:2],
  method = "Weighted Average", 
  mukeys = 341081,
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)

```

**Comment by brownag** [comment]
Thanks for reporting this and continuing to stress test these functions!

This is definitely unexpected behavior--and and edge case. 1/10bar water content is not populated for any of the layers occurring within [0,25]... and somehow that is messing up the results for all properties. The other variables should be independent of the assessment of missing data in `wtenthbar_r`, but clearly that is not the case...

Either increasing the bottom depth so at least one  `wtenthbar_r`  value is included, or excluding `wtenthbar_r` appears to give expected, non-NA results.

**Comment by brownag** [comment]
Here is a reprex based on the new `slab()` function [see draft PR https://github.com/ncss-tech/aqp/pull/259] supporting site-level weights. It show values I am trying to reproduce with `method="Weighted Average"`. The target values must be able to be determined simultaneously in a single call to `get_SDA_property()` 

Narrowing in on the problem: I got it working for the [0, 25] case but somewhere the weighting is still slightly off in [0, 100]

``` r
library(aqp)
#> This is aqp 1.43
library(soilDB)
vars <- c("wtenthbar_r", "ph1to1h2o_r")
x <- get_SDA_property(
  property = vars,
  method = "none",
  mukeys = 341081,
  include_minors = TRUE,
  miscellaneous_areas = FALSE
)
depths(x) <- cokey ~ hzdept_r + hzdepb_r
#> converting profile IDs from integer to character
site(x) <- ~ mukey + comppct_r

# [0,25]
slab(
  x,
  mukey ~ wtenthbar_r + ph1to1h2o_r,
  slab.structure = c(0, 25),
  weights = "comppct_r",
  slab.fun = weighted.mean,
  na.rm = TRUE
)
#>      variable  mukey   value contributing_fraction top bottom
#> 1 wtenthbar_r 341081     NaN                     0   0     25
#> 2 ph1to1h2o_r 341081 8.14696                     1   0     25

# [0,100]
slab(
  x,
  mukey ~ wtenthbar_r + ph1to1h2o_r,
  slab.structure = c(0, 100),
  weights = "comppct_r",
  slab.fun = weighted.mean,
  na.rm = TRUE
)
#>      variable  mukey    value contributing_fraction top bottom
#> 1 wtenthbar_r 341081 10.78891                 0.198   0    100
#> 2 ph1to1h2o_r 341081  8.15776                 1.000   0    100
```

**Comment by brownag** [comment]
Close
```r
library(aqp)
library(soilDB)
vars <- c("wtenthbar_r", "ph1to1h2o_r")

x <- get_SDA_property(
  property = vars,
  method = "none",
  mukeys = 341081,
  include_minors = TRUE,
  miscellaneous_areas = FALSE
)

depths(x) <- cokey ~ hzdept_r + hzdepb_r
site(x) <- ~ mukey + comppct_r

x |> 
  trunc(0, 100) |> 
  mutate_profile(w1 = weighted.mean(wtenthbar_r, hzdepb_r - hzdept_r, na.rm = TRUE),
                 w2 = weighted.mean(ph1to1h2o_r, hzdepb_r - hzdept_r, na.rm = TRUE)) |> 
  site() |> 
  (\(x) data.frame(wtenthbar_r = signif(weighted.mean(x$w1, x$comppct_r, na.rm = TRUE), 3),
                   ph1to1h2o_r = signif(weighted.mean(x$w2, x$comppct_r, na.rm = TRUE), 3)))(x = _)

get_SDA_property(
  property = vars,
  method = "weighted average",
  mukeys = 341081,
  include_minors = TRUE,
  miscellaneous_areas = FALSE
)
```

**Comment by dylanbeaudette** [comment]
I just tested after that latest commit, and it seems to have fixed a similar bug with different combinations of properties.

```r
library(soilDB)

# worked
v <- c("sandtotal_r", "silttotal_r", "claytotal_r")
.mukey <- 565473

get_SDA_property(
  property = v,
  method = "Weighted Average", 
  mukeys = as.integer(.mukey),
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)


# did not work prior to latest commit
# works now
v <- c("sandtotal_r", "silttotal_r", "claytotal_r", "om_r", "cec7_r", "ph1to1h2o_r", "wthirdbar_r", "wfifteenbar_r")
get_SDA_property(
  property = v,
  method = "Weighted Average", 
  mukeys = as.integer(.mukey),
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)
```

**Comment by brownag** [comment]
Ok, good. 

Yeah I think cec7 will suffer from similar problems as wtenthbar since they both are routinely not populated due to specific decisions made in the standard NASIS calculations. 

This was a nasty bug...really obvious with above two mentioned properties but something more subtly affecting the more commonly populated properties. 

Perhaps a good reason to move forward with porting all the TSQL specific functionality to something a bit more robust,  ideally I would also like SQLite compatibility out of it, and with all the temp tables required in current implementation that is difficult. 

I have some ideas on a refactor but hopefully above fix works for now. I still want to investigate some of the rounding and possible floating point errors accumulated by this method.

**Comment by dylanbeaudette** [comment]
A new development, there are some set of map unit keys that result in an SDA error. At first I thought that this was a resource constraint (query time to number of rows), however, it works fine with no aggregation.

```r
library(soilDB)


v <- c("sandtotal_r", "silttotal_r", "claytotal_r", "ph1to1h2o_r", "wthirdbar_r", "wfifteenbar_r")

.mukey <- c(1949850, 2586191, 2586192, 2586194, 2586195, 2586200, 2586204, 
            2586209, 2586211, 2586212, 2586213, 2586219, 2586224, 2586225, 
            2586229, 2587031, 2587035, 2587037, 2587038, 2587042, 2587044, 
            2587045, 2587047, 2587049, 2587055, 2587056, 2587058, 2587059, 
            2587061, 2587065, 2587501, 2587502, 2587503, 2587504, 2587505, 
            2587506, 2587509, 2587511, 2587514, 2587515, 2587516, 2587517, 
            2587518, 2587519, 2587520, 2587521, 2587522, 2587523, 2587524, 
            2587525, 2587526, 2587527, 2587528, 2587529, 2587530, 2587531, 
            2587533, 2587534, 2587535, 2587537, 2587538, 2587539, 2587541, 
            2587542, 2587543, 2587546, 2587547, 2587549, 2587550, 2587551, 
            2587552, 2587553, 2587554, 2587555, 2587556, 2587557, 2587558, 
            2587559, 2587561, 2587562, 2587563, 2587565, 2587566, 2587567, 
            2587568, 2587569, 2587570, 2587574, 2587575, 2587576, 2587578, 
            2587579, 2587580, 2587583, 2587584, 2587585, 2587588, 2587589, 
            2587590, 2587591, 2587994, 2588007, 2588028, 2588036, 2588073, 
            2588085, 2588095, 2588099, 2588101, 2588110, 2588119, 2588130, 
            2588141, 2588144, 2588152, 2588186, 2588189, 2588199, 2588217, 
            2588225, 2588230, 2588234, 2624631, 2624632, 2624633, 2642946, 
            2642947, 2642948, 2642950, 2642951, 2642952, 2642953, 2642954, 
            2642955, 2642956, 2642957, 2642958, 2642959, 2642960, 2642962, 
            2642963, 2642964, 2642965, 2642966, 2642967, 2642971, 2642972, 
            2642973, 2642974, 2642975, 2642976, 2642978, 2642984, 2642985, 
            2642988, 2642989, 2642994, 2642996, 2642997, 2642998, 2643001, 
            2643002, 2643004, 2643005, 2643008, 2643009, 2643010, 2643015, 
            2643018, 2643023, 2643024, 2643027, 2643028, 2643033, 2643034, 
            2643035, 2643036, 2643038, 2643039, 2643040, 2643041, 2643042, 
            2643043, 2643044, 2643045, 2643046, 2643047, 2643063, 2643064, 
            2643067, 2643068, 2643072, 2643073, 2643075, 2643076, 2643078, 
            2643082, 2643083, 2643085, 2643087, 2643089, 2643091, 2643093, 
            2643094, 2643097, 2643098, 2643099, 2643100, 2643101, 2643103, 
            2643106, 2643107, 2798405, 2798406, 2798407, 2798409, 2798410, 
            2798411, 2798412, 2798413, 2798415, 2798417, 2798418, 2798419, 
            2798420, 2798422, 2798423, 2798424, 2798426, 2798427, 2798429, 
            2798830, 2798831, 2798832, 2798833, 2798834, 2798836, 2798837, 
            2798838, 2798839, 2798840, 2798841, 2798842, 2798843, 2798844, 
            2798845, 2798846, 2798847, 2798850, 2798851, 2798852, 2798853, 
            2798854, 2798855, 2798857, 2798858, 2798859, 2798860, 2798861, 
            2798862, 2798863, 2798864, 2798866, 2798867, 2798873, 2798875, 
            2798876, 2798877, 2798878, 2798879, 2798880, 2798882, 2798883, 
            2798884, 2798885, 2798886, 2798887, 2798888, 2798890, 2903032, 
            3184687, 3184688, 3184689, 3184690, 3184691, 3184692, 3184693, 
            3184694, 3184695, 3184696, 3184697, 3184698, 3184699, 3184700, 
            3184701, 3184702, 3184703, 3247630, 3247631, 3247634, 3247752
)

# SDA error
x <- get_SDA_property(
  property = v,
  method = "Weighted Average", 
  mukeys = as.integer(.mukey),
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)

# works as expected
x <- get_SDA_property(
  property = v,
  method = "none", 
  mukeys = as.integer(.mukey),
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)
```

**Comment by brownag** [comment]
Great find; an error was due to division by zero in some cases--related to recent changes and NA values for sand/silt/clay for the mucky surface in your problem mapunit. As above the expected values for those properties are NA and the presence of NA should not affect the other properties.

It appears that was the only problem here, but some of the other mapunits might have same issue. Should now be fixed in 55cb916

``` r
library(soilDB)


v <- c("sandtotal_r", "silttotal_r", "claytotal_r", "ph1to1h2o_r", "wthirdbar_r", "wfifteenbar_r")

.mukey <- c(2588095)

get_SDA_property(
  property = v,
  method = "Weighted Average", 
  mukeys = as.integer(.mukey),
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)
#>   areasymbol musym                                            muname   mukey
#> 1      ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#>   sandtotal_r silttotal_r claytotal_r ph1to1h2o_r wthirdbar_r wfifteenbar_r
#> 1          NA          NA          NA         7.1       45.18         13.63

get_SDA_property(
  property = v,
  method = "none", 
  mukeys = as.integer(.mukey),
  top_depth = 0,
  bottom_depth = 25,
  include_minors = TRUE, 
  miscellaneous_areas = FALSE
)
#>    areasymbol musym                                            muname   mukey
#> 1       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 2       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 3       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 4       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 5       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 6       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 7       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 8       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 9       ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#> 10      ND079 F442A Eramosh mucky peat, ponded, 0 to 1 percent slopes 2588095
#>    compname compkind comppct_r majcompflag    cokey    chkey hzdept_r hzdepb_r
#> 1   Eramosh   Series        77         Yes 21562262 63579928        0       13
#> 2   Eramosh   Series        77         Yes 21562262 63579927       13       25
#> 3   Eramosh   Series        77         Yes 21562262 63579930       25       51
#> 4   Eramosh   Series        77         Yes 21562262 63579931       51      115
#> 5   Eramosh   Series        77         Yes 21562262 63579929      115      200
#> 6   Eramosh   Series        17         No  21562263 63579933        0       13
#> 7   Eramosh   Series        17         No  21562263 63579932       13       25
#> 8   Eramosh   Series        17         No  21562263 63579935       25       51
#> 9   Eramosh   Series        17         No  21562263 63579934       51      115
#> 10  Eramosh   Series        17         No  21562263 63579936      115      200
#>    sandtotal_r silttotal_r claytotal_r ph1to1h2o_r wthirdbar_r wfifteenbar_r
#> 1           NA          NA          NA         7.2        41.2          14.3
#> 2           NA          NA          NA         7.0        49.5          12.9
#> 3           15          60          25         7.9        33.8          21.5
#> 4           12          60          28         7.9        33.6          21.1
#> 5            5          60          35         7.9        34.1          22.0
#> 6           NA          NA          NA         7.2        41.2          14.3
#> 7           NA          NA          NA         7.0        49.5          12.9
#> 8           15          60          25         7.9        33.8          21.5
#> 9           12          60          28         7.9        33.6          21.1
#> 10           5          60          35         7.9        34.1          22.0
```

**Comment by dylanbeaudette** [comment]
Awesome, thanks! Testing now.
